tests/unit/live/test_rbtree.c

```diff
@@ -86,6 +88,127 @@ void populate_int_arrays()
+{
+  TEST_ASSERT_NOT_NULL(etcpal_rbtree_init(self, int_compare, node_alloc, node_dealloc));
+
+  // Insert even values between 0 and 98 (inclusive) into the tree at random.
+  for (int i = 0; i < INT_ARRAY_SIZE; ++i)
+  {
+    if ((random_int_array[i] % 2) == 0)
+    {
+      TEST_ASSERT_EQUAL(kEtcPalErrOk, etcpal_rbtree_insert(self, &random_int_array[i]));
+    }
+  }
+}
+
+/* Pass in 0 to test upper_bound, or 1 to test lower_bound. */
+void test_bound(int is_lower_bound)
+{
+  // Initialize the tree and iterator. Randomly initialize the tree with even numbers between 0 and 98 inclusive.
+  EtcPalRbTree tree;
+  EtcPalRbIter iter;
+  initialize_tree_with_random_even_ints(&tree);
+  etcpal_rbiter_init(&iter);
+
+  const char* function_name_log_str = (is_lower_bound ? "lower_bound" : "upper_bound");
+
+  // Test bound of ints from 0 to 99 inclusive.
+  for (int i = 0; i < INT_ARRAY_SIZE; ++i)
+  {
+    int expected_value = -1;
+
+    if (is_lower_bound)
+    {
+      // Even numbers should have equivalent lower bounds.
+      // The lower bounds of the odd numbers should be the even number immediately after (if any).
+      expected_value = ((i % 2) == 0) ? (i) : (i + 1);
+    }
+    else
+    {
+      // The upper_bound should be the even number immediately after.
+      expected_value = ((i % 2) == 0) ? (i + 2) : (i + 1);
+    }
+
+    int* bound = (int*)(is_lower_bound ? etcpal_rbiter_lower_bound(&iter, &tree, &i)
+                                       : etcpal_rbiter_upper_bound(&iter, &tree, &i));
+
+    // Check that this value can be found with etcpal_rbtree_find.
+    if (bound)
+    {
+      int* found = (int*)etcpal_rbtree_find(&tree, bound);
+
+      char test_null_error_msg[100];
+      sprintf(test_null_error_msg, "The %s function for %d returned %d, which isn't in the red-black tree.",
+              function_name_log_str, i, *bound);
+      TEST_ASSERT_NOT_NULL_MESSAGE(found, test_null_error_msg);
+    }
+
+    // Check the return value against the expected result.
+    if (expected_value >= INT_ARRAY_SIZE)  // Expect NULL.
+    {
+      int test_error_value = -1;
+      if (bound)
+      {
+        test_error_value = *bound;
+      }
+
+      char test_not_null_error_msg[100];
+      sprintf(test_not_null_error_msg, "The %s function for %d returned %d, when it should have returned NULL.",
+              function_name_log_str, i, test_error_value);
+
+      TEST_ASSERT_NULL_MESSAGE(bound, test_not_null_error_msg);
+    }
+    else  // Expect non-NULL.
+    {
+      char test_null_error_msg[100];
+      sprintf(test_null_error_msg, "The %s function for %d returned NULL, when it should have returned %d.",
+              function_name_log_str, i, expected_value);
+      TEST_ASSERT_NOT_NULL_MESSAGE(bound, test_null_error_msg);
+
+      if (bound)
+      {
+        char test_not_equal_error_msg[100];
+        sprintf(test_not_equal_error_msg, "The %s function for %d returned %d, when it should have returned %d.",
```

Comment:
Same comment as above; also, printing the values of *bound and expected_value here is duplicating functionality that the unit test framework gives us. If the test fails, it will print the two values and the fact that they are not equal.